### PR TITLE
FEA: Function line numbers track point of use and point of expansion

### PR DESCRIPTION
**Pull request by daz10000**
Hi Chris,  this is the fix for the function line numbers.  Check it out carefully as it's changing the core Node type.  I have tried to make it backward compatible but also provide the extended functionality with multiple tracked locations.  

* Zero argument functions are implemented and tested via the line number tracking tests.

* Tests provided and pass.   

* In the multiply recursive test example, I am only getting first and last parts of expansion (point of initial call and point of assembly generation) in the line numbers.  I think the expansion is expanding all the internal functions before I get to them.  If you have any suggestions for improving this let me know.  It implements what I need right now.

* There's a stray primercreation update in here that implements the Tm constrain.  I swear that was already in the code base but I suspect it hasn't been merged yet.  It was in my master branch unfortunately and if you are ok blessing that, it will make my life a lot easier ;( - it's getting hard to maintain a clean master branch and also port changes into our production branch.

### CODE REVIEW

**Comment by daz10000** [comment]
I cleaned up a few small merge conflicts due to version bump.   You should consider increasing version number as part of the release as well of course but that's your call.

File: src/GslCore/AstProcess.fs [hunk 1]
```diff
@@ -326,19 +326,44 @@ let private inlinePassedArgs (vb: VariableBindings) (fd, fc: FunctionCall) =
...
-        | Some(fd) ->
-            // inline the args into the function call block
-            // this new block replaces the function call
-            checkArgs fd fc node
-            >>= inlinePassedArgs s.vars
-        | None -> error UnresolvedFunction fc.name node
+
+        // step back one character since tokenizer is already past end when snapshot was token
+        // aim to point at last character of function name.
+        // let e = { fcw.Pos.Value.e with pos_cnum = fcw.Pos.Value.e.pos_cnum-1 }
```
Comment (chrismacklin) [inline src/GslCore/AstProcess.fs]: Is this now a stale comment and line of code?

File: src/GslCore/AstTypes.fs [hunk 2]
```diff
@@ -90,28 +90,36 @@ let posBracketTokens left right : SourcePosition = {s = left.pos.s; e = right.po
 /// Wrapper type for every AST node.
 /// This enables adding extensible metadata to the AST for tracking things such as source code position.
 [<CustomEquality>][<NoComparison>]
-type Node<'T when 'T: equality> = {x: 'T; pos: SourcePosition option}
+type Node<'T when 'T: equality> = {x: 'T; pos: SourcePosition list} // most recent position at head
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: Three requests:
1. please move the descriptive comment here into the type declaration as a docstring comment on the record field.
2. I don't quite understand what "most recent position" means in this context.  Could you rephrase to clarify?
3. Now that we have both singular and collections of positions floating around, I think we need to do some variable renaming.  Here I think we should replace `pos` to be `positions`.  Having `x.pos` return a list while `x.Pos` returns an option is a confusing API.

File: src/GslCore/AstTypes.fs [hunk 3]
```diff
@@ -216,7 +224,80 @@ and AstNode =
     | Splice of AstNode []
     with
     /// Get the position from a node.
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: This docstring needs to be updated to describe this method.

File: src/GslCore/AstTypes.fs [hunk 4]
```diff
@@ -461,8 +543,13 @@ let posFromList (nodes: AstNode list) : SourcePosition option =
...
 let createParseError msg pos = ParseError({x=msg; pos=pos})
 
+let posOptionToList (pos:SourcePosition option) =
+    match pos with
+    | Some p -> [p]
+    | None -> []
+
 /// Wrap a value with position taken from another node.
-let nodeWrapWithNodePosition (node: AstNode) v = {x = v; pos = node.pos}
+let nodeWrapWithNodePosition (node: AstNode) v = {x = v; pos = node.pos |> posOptionToList}
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: Is this the right thing to do here?  Seems like we should absorb the entire list of positions from the node from which we are deriving position.

File: src/GslCore/AstTypes.fs [hunk 5]
```diff
@@ -461,8 +543,13 @@ let posFromList (nodes: AstNode list) : SourcePosition option =
 /// Create a parse error with message and position.
 let createParseError msg pos = ParseError({x=msg; pos=pos})
 
+let posOptionToList (pos:SourcePosition option) =
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: This function duplicates the builtin `Option.toList`; please replace with it.

File: src/GslCore/AstTypes.fs [hunk 6]
```diff
@@ -216,7 +224,80 @@ and AstNode =
...
+        | ParsePragma(nw) -> nw.Pos
+        | Pragma(nw) -> nw.Pos
+        | Block(nw) -> nw.Pos
+        | FunctionDef(nw) -> nw.Pos
+        | FunctionLocals(nw) -> nw.Pos
+        | FunctionCall(nw) -> nw.Pos
+        | Assembly(nw) -> nw.Pos
+        | ParseError(nw) -> nw.Pos
+        | Splice(_) -> None
+    member x.positions =
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: Let's trim some source code here.  We only need to define `x.positions`, and then we can define the above `x.pos` property as
```
member x.pos = x.positions |> List.tryHead
```

File: src/GslCore/AstTypes.fs [hunk 7]
```diff
@@ -551,19 +638,19 @@ let createMutation (s: PString) mutType =
     let t = mutStr.[mutStr.Length-1]
     let pos = Convert.ToInt32(mutStr.[2..mutStr.Length-2])
     let mut = {f = f; t = t; loc = pos; mType = mutType}
-    Mutation({x = mut; pos = Some s.pos})
+    Mutation({x = mut; pos = [s.pos]})
 
 /// Create a top-level part.
 let createPart mods pragmas basePart =
-    Part({x = {basePart = basePart; mods = mods; pragmas = pragmas; fwd = true}; pos = basePart.pos})
+    Part({x = {basePart = basePart; mods = mods; pragmas = pragmas; fwd = true}; pos = basePart.pos |> posOptionToList})
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: This is another case where I think we should take the entire position list rather than just the (possible) head.

File: src/GslCore/PluginTypes.fs [hunk 8]
```diff
@@ -131,16 +132,16 @@ type AssemblyTransformationMessage<'A when 'A :> ISourcePosition> =
         let verbose = defaultArg verbose false
         seq {
             match (x.assembly :> ISourcePosition).OptionalSourcePosition with
-            | Some(p) ->
-                yield sprintf "%O during %s %s:" x.kind phase (p.Format())
+            | [] ->
+                yield sprintf "%O during %s:" x.kind phase
+                yield x.msg
+            |  hd::tl -> 
+                yield sprintf "%O during %s %s:" x.kind phase (String.Join(";",hd::tl |> List.map (fun p ->p.Format())))
```
Comment (chrismacklin) [inline src/GslCore/PluginTypes.fs]: Could you use "; " as the separator to improve readability?

File: tests/GslCore.Tests/TestLineNumbers.fs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                match p.x.basePart with
+                | Assembly a as x -> yield x
+                | _ -> ()
+            | Assembly a as x -> yield x
+            | _ -> ()
+        ]
+
+    /// NB: note lines are numbered from zero internally, and columns.
+    /// The parser typically reports one character past the end of the identifier
+    let checkPosition (pos:SourcePosition option) (startLine,startCol,endLine,endCol) =
```
Comment (chrismacklin) [inline tests/GslCore.Tests/TestLineNumbers.fs]: Please follow each comma with a space.

File: tests/GslCore.Tests/TestLineNumbers.fs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+end
+
+
+knockout(uADH1,dADH1)
+knockout(uADH2,dADH2)
+knockout(uADH3,dADH3)
+knockout(uADH4,dADH4)""" 
+
+        let result = 
+            source
```
Comment (chrismacklin) [inline tests/GslCore.Tests/TestLineNumbers.fs]: It would be nice to extract this pipeline out as it is used in every test.

File: src/GslCore/AstProcess.fs [hunk 11]
```diff
@@ -326,19 +326,44 @@ let private inlinePassedArgs (vb: VariableBindings) (fd, fc: FunctionCall) =
         | _ -> error (InternalError(TypeError)) "No function locals node found in function defintion block." fd.body
     | x -> internalTypeMismatch (Some "function body") "Block" x
 
+let private renumberLines (newPos:SourcePosition) (block:AstNode) =
+    let rec updatePos (newPos:SourcePosition) (node:AstNode) =
+        match node with
+        | Block b -> Block {b with x = b.x |> List.map (updatePos newPos) }
+        | Splice s -> Splice (s |> Array.map (updatePos newPos))
```
Comment (chrismacklin) [inline src/GslCore/AstProcess.fs]: This line isn't necessary; line renumbering in AST splices is handled by the literal source bootstrapping mechanism already in AstExpansion.fs.

File: src/GslCore/AstProcess.fs [hunk 12]
```diff
@@ -326,19 +326,44 @@ let private inlinePassedArgs (vb: VariableBindings) (fd, fc: FunctionCall) =
...
+        | Part p ->  Part {p with x = { p.x with basePart = updatePos newPos p.x.basePart}}
```
Comment (chrismacklin) [inline src/GslCore/AstProcess.fs]: It strikes me that the correct thing to do here is actually simpler than what's going on.  We already have a function that adds a position to an AST node and returns an updated node - that's `AstNode.PushPos`, though it will need to be refactored into a function rather than a method.  We also have a function, `map`, that produces an updated AST by mapping a function.  So I think we don't actually need a `renumberLines` function.  See below comment for a suggested refactoring.

File: src/GslCore/AstProcess.fs [hunk 13]
```diff
@@ -326,19 +326,44 @@ let private inlinePassedArgs (vb: VariableBindings) (fd, fc: FunctionCall) =
...
+        | None ->
+            error (InternalError(TypeError)) "Expected function call to have a source position" node
+        | Some newPos ->
+            match s.defs.TryFind(fc.name) with
+            | Some(fd) ->
+                // inline the args into the function call block
+                // this new block replaces the function call
+                checkArgs fd fc node
+                >>= inlinePassedArgs s.vars
+                >>= renumberLines {newPos with e = e}
```
Comment (chrismacklin) [inline src/GslCore/AstProcess.fs]: Instead of `renumberLines`, I think this will do the trick:

In AstTypes.fs, refactor `AstNode.PushPos` into a function of signature `SourcePosition -> AstNode -> AstNode`; perhaps call it `pushPos`.

Now, here:
```
let addPos = pushPos newPos

checkArgs fd fc node
>>= inlinePassedArgs s.vars
>>= renumberLines {newPos with e = e}
|> lift AstTreeHead // needed to adapt to the map function
>>= map Serial TopDown addPos
|> lift (fun treeHead -> treeHead.wrappedNode)
```

Comment (chrismacklin) [reply src/GslCore/AstProcess.fs]: I *think* this refactoring may also fix the fact that nested function calls don't get all of the expected positions, but I'm not sure.

**Review by chrismacklin — CHANGES_REQUESTED** [review]
Great improvement overall, and I think with these small fixes we can get it all the way there.

File: src/GslCore/AstProcess.fs (hunk 1 above)
Comment (daz10000) [reply src/GslCore/AstProcess.fs]: yes. fixed.

File: src/GslCore/AstTypes.fs (hunk 2 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: All good points. Moved doc string,  improved it and I like positions, so made the 156 changes where it was used.  And it still compiles!

File: src/GslCore/AstTypes.fs (hunk 3 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Updated .pos.  Fixed PushPos.  Added docstring too for .positions.

File: src/GslCore/AstTypes.fs (hunk 4 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Good point.  Was hoping this would fix the problem with multiply recursive function expansion losing the middle positions but did not,.  Implemented anyway,

File: src/GslCore/AstTypes.fs (hunk 5 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: There's a lot packed into the Option space!.  Done.

File: tests/GslCore.Tests/TestLineNumbers.fs (hunk 10 above)
Comment (daz10000) [reply tests/GslCore.Tests/TestLineNumbers.fs]: Good idea.  Done.

File: src/GslCore/AstProcess.fs (hunk 11 above)
Comment (daz10000) [reply src/GslCore/AstProcess.fs]: Removed.

File: src/GslCore/AstProcess.fs (hunk 13 above)
Comment (daz10000) [reply src/GslCore/AstProcess.fs]: So basic idea works well (though didn't fix the nested function problem alas).   renumberLines is gone.

File: src/GslCore/PluginTypes.fs (hunk 8 above)
Comment (daz10000) [reply src/GslCore/PluginTypes.fs]: You mean between the different locations? I think it is using the semi colon now if I'm reading it correctly.  Might have just fixed that inadvertently but let me know if I missed something.

File: tests/GslCore.Tests/TestLineNumbers.fs (hunk 9 above)
Comment (daz10000) [reply tests/GslCore.Tests/TestLineNumbers.fs]: done.

File: src/GslCore/AstProcess.fs (hunk 12 above)
Comment (daz10000) [reply src/GslCore/AstProcess.fs]: implemented.

File: src/GslCore/PluginTypes.fs (hunk 8 above)
Comment (chrismacklin) [reply src/GslCore/PluginTypes.fs]: Sorry that may have been hard to read; I was suggesting a semicolon followed by a space, so the resulting message reads more like a sentence.

File: tests/GslCore.Tests/TestLineNumbers.fs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+        let a1 = List.head assemblies
+        checkPosition a1.pos (11,0,11,4)
+
+        // test coordinates of second assembly
+        let a2 = List.item 1 assemblies
+        checkPosition a2.pos (12,0,12,6)
+
+        // test coordinates of last assembly
+        let a3 = List.item 2 assemblies
+        checkPosition a3.pos (19,9,19,22)
```
Comment (chrismacklin) [inline tests/GslCore.Tests/TestLineNumbers.fs]: I'm assuming that this is the test that would catch the issue with missing nested function call positions?  Would love to fix this in the future.  Could you do the following:

1. File an issue that documents the problem, maybe with a GSL snippet that reproduces it.
2. Add either a failing but skipped unit test, or some commented-out assertions in this test, that would pass if we fixed this problem.
3. Cross-reference the issue with the test.

File: src/GslCore/PluginTypes.fs (hunk 8 above)
Comment (daz10000) [reply src/GslCore/PluginTypes.fs]: ah - got it.  Done.

File: tests/GslCore.Tests/TestLineNumbers.fs (hunk 14 above)
Comment (daz10000) [reply tests/GslCore.Tests/TestLineNumbers.fs]: Done.  See issue #15 .  There is a test marked ignore that asserts correct behavior and currently fails.

File: src/GslCore/AstTypes.fs (hunk 6 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Nice catch.  Done.

File: src/GslCore/AstTypes.fs (hunk 7 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Done.  This (again) might be cause of the missing line numbers.

**Review by daz10000 — COMMENTED** [review]
I think I've addressed the outstanding change requests but let me know.

File: src/GslCore/AstProcess.fs [hunk 15]
```diff
@@ -326,19 +326,31 @@ let private inlinePassedArgs (vb: VariableBindings) (fd, fc: FunctionCall) =
...
         let fc = fcw.x
-        match s.defs.TryFind(fc.name) with
-        | Some(fd) ->
-            // inline the args into the function call block
-            // this new block replaces the function call
-            checkArgs fd fc node
-            >>= inlinePassedArgs s.vars
-        | None -> error UnresolvedFunction fc.name node
+
+        match fcw.Pos with
```
Comment (chrismacklin) [inline src/GslCore/AstProcess.fs]: Hmm; why is this match on the optionalized version of the source position?  Don't we want to take the entire list and concatenate it with that on the node rather than pushing just the head?  Seems like this *might* actually be the source of the missing intermediate call sites?  Changing this to `fcw.positions` (and possibly removing the branch that would fail if the list were empty which currently catches None) will probably get us there!

File: src/GslCore/AstTypes.fs [hunk 16]
```diff
@@ -88,30 +88,43 @@ let posBracketTokens left right : SourcePosition = {s = left.pos.s; e = right.po
...
 /// Wrapper type for every AST node.
-/// This enables adding extensible metadata to the AST for tracking things such as source code position.
+/// This enables adding extensible metadata to the AST for tracking things such as source 
+/// code position.  pos tracks position history of this node (when functions expand) with highest level call first
+/// with 
 [<CustomEquality>][<NoComparison>]
-type Node<'T when 'T: equality> = {x: 'T; pos: SourcePosition option}
+type Node<'T when 'T: equality> = {x: 'T; positions: SourcePosition list}
     with
+    member x.Pos =
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: This property might be unused after refactoring the position addition code to use the full list of positions?  Not sure though.

File: src/GslCore/AstTypes.fs [hunk 17]
```diff
@@ -215,48 +228,54 @@ and AstNode =
     // to explode them into their outer contexts.
     | Splice of AstNode []
     with
-    /// Get the position from a node.
-    member x.pos =
+    /// Add a new position to this node.  This occurs during function expansion when an instance
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: This docstring appears to refer to a different (possibly now deleted?) method?

File: src/GslCore/LegacyParseTypes.fs [hunk 18]
```diff
@@ -189,7 +189,7 @@ let private createLegacyPart part =
     | Gene(gw) ->
         convertMods part.x.mods
         >>= (fun mods ->
-            let where = match gw.pos with | Some(p) -> p | None -> emptySourcePosition
+            let where = match gw.positions with | hd::_tl -> hd | [] -> emptySourcePosition
             let genePart = {gene = gw.x.gene; mods = mods; where = where}
```
Comment (chrismacklin) [inline src/GslCore/LegacyParseTypes.fs]: Not a blocker, but at some point we may want to refactor these downstream types to maintain lists of positions as well.

File: src/GslCore/PluginTypes.fs [hunk 19]
```diff
@@ -1,5 +1,6 @@
 ﻿/// Definitions of plug-in types and interfaces.
 module PluginTypes
+open System
```
Comment (chrismacklin) [inline src/GslCore/PluginTypes.fs]: Is this import actually used?

**Review by chrismacklin — APPROVED** [review]

File: src/GslCore/AstProcess.fs (hunk 15 above)
Comment (daz10000) [reply src/GslCore/AstProcess.fs]: Good call - that fixes the missing line number problem.  I have re-enabled the nested call test and made it more detailed (checks number and position of values).

File: src/GslCore/AstTypes.fs (hunk 16 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Again good observation backed up by experiemental evidence.  Deleted.

File: src/GslCore/AstTypes.fs (hunk 17 above)
Comment (daz10000) [reply src/GslCore/AstTypes.fs]: Yep - removed

File: src/GslCore/PluginTypes.fs (hunk 19 above)
Comment (daz10000) [reply src/GslCore/PluginTypes.fs]: Yes - needed for String.Join later.

Comment (daz10000) [reply src/GslCore/PluginTypes.fs]: Actually just managed to get rid of it.

File: src/GslCore/LegacyParseTypes.fs (hunk 18 above)
Comment (daz10000) [reply src/GslCore/LegacyParseTypes.fs]: Fixed GenePart.  Centralized the formatting too for reporting multiple locations, which removed the need for the system import below too

**Comment by daz10000** [comment]
Ok, one more round of cleaning and a bug fix and some slightly improved functionality.

File: src/GslCore/AstTypes.fs [hunk 20]
```diff
@@ -352,6 +367,45 @@ and Roughage = {locus: Node<L2Id> option; marker: Node<string> option; parts: No
             | Some(mw) -> Some(mw.x)
         | Some(mw) -> Some(mw.x) // Yes there is a marker attached to the locus knockout
 
+/// Helper function for pushing new source code positions onto stack during function expansion
+/// Most recently (first) postition refers to highest level function invocation.
+let pushPos (newPos:SourcePosition) (x:AstNode) =
```
Comment (chrismacklin) [inline src/GslCore/AstTypes.fs]: I think this should be refactored to to prepend an entire new list of positions; that will simplify the sole use of this code in the function expansion section.  Basically pushPosNode becomes, say, prependPositionsNode:
```
let prependPositionsNode (newPositions: SourcePosition list) (node: Node<'T>) =
    {node with positions=newPositions@node.positions}
```
Then this function gains the signature `SourcePosition list -> AstNode -> AstNode` and the `List.rev` and `List.fold` in the function renumbering function should disappear.  I *think* this gives us the correct order including nested positions.
